File: Leetcode100/leetcode100_python/SlidingWindowMaximum.py
from typing import List

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        result = []
        deque = []
        for i in range(n):
            if deque and deque[0] <= i - k:
                deque.pop(0)
            while deque and nums[deque[-1]] < nums[i]:
                deque.pop()
            deque.append(i)
            if i >= k - 1:
                result.append(nums[deque[0]])
        return result

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    k = int(input())
    solution = Solution()
    print(solution.maxSlidingWindow(nums, k))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/sliding_window_maximum/SlidingWindowMaximum.java
package main.java.com.jiang.leetcode100.sliding_window_maximum;

import java.util.PriorityQueue;
import java.util.Scanner;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        int n = nums.length;
        int[] result = new int[n - k + 1];

        int max_num = Integer.MIN_VALUE;
        int index = 0;

        for(int i = 0; i < k; i++) {
            max_num = Math.max(max_num, nums[i]);
        }
        result[index++] = max_num;

        for(int i = k; i < n; i++) {
            if(nums[i] > max_num) {
                max_num = nums[i];
            } else if(nums[i - k] == max_num) {
                max_num = Integer.MIN_VALUE;
                for(int j = i - k + 1; j <= i; j++) max_num = Math.max(nums[j], max_num);
            }
            result[index++] = max_num;
        }

        return result;
    }
}

public class SlidingWindowMaximum {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        int k = scanner.nextInt();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        int[] result = new Solution().maxSlidingWindow(nums, k);
        for(int i = 0; i < result.length; i++) System.out.print(result[i] + " ");
    }
}